File: frontend/src/utils/logger.js
/**
 * Structured frontend logger with module tagging.
 */

const emit = (level, module, event, data = {}) => {
    const payload = {
        timestamp: new Date().toISOString(),
        level,
        module,
        event,
        data,
    };

    if (level === 'error') {
        console.error(payload);
        return;
    }
    if (level === 'warn') {
        console.warn(payload);
        return;
    }
    console.info(payload);
};

export const createLogger = (module) => ({
    info: (event, data) => emit('info', module, event, data),
    warn: (event, data) => emit('warn', module, event, data),
    error: (event, data) => emit('error', module, event, data),
});

File: backend/utils/logger.py
"""Structured logger utility with module tagging and timestamps."""

from __future__ import annotations

import json
import logging
from datetime import datetime, timezone
from typing import Any


class AppLogger:
    """Small structured logger wrapper used by service modules."""

    def __init__(self, module: str) -> None:
        self.module = module
        self._logger = logging.getLogger(f"greengate.{module}")

    def _emit(self, level: int, event: str, data: dict[str, Any] | None = None) -> None:
        payload = {
            "timestamp": datetime.now(timezone.utc).isoformat(),
            "module": self.module,
            "event": event,
            "data": data or {},
        }
        self._logger.log(level, json.dumps(payload, default=str))

    def info(self, event: str, data: dict[str, Any] | None = None) -> None:
        self._emit(logging.INFO, event, data)

    def warn(self, event: str, data: dict[str, Any] | None = None) -> None:
        self._emit(logging.WARNING, event, data)

    def error(self, event: str, data: dict[str, Any] | None = None) -> None:
        self._emit(logging.ERROR, event, data)


def get_logger(module: str) -> AppLogger:
    """Return a structured logger scoped to a module."""
    return AppLogger(module)
